File: sample/src/main/java/com/smartnsoft/smartrecyclerviewsample/adapter/SampleDiffUtilSmartRecyclerAdapter.java
package com.smartnsoft.smartrecyclerviewsample.adapter;

import java.util.List;

import android.content.Context;

import com.smartnsoft.recyclerview.adapter.DiffUtilSmartRecyclerAdapter;
import com.smartnsoft.recyclerview.attributes.SmartRecyclerAttributes;
import com.smartnsoft.recyclerview.wrapper.SmartRecyclerViewWrapper;

/**
 * A simple implementation of the DiffUtilSmartRecyclerAdapter
 *
 * @author Adrien Vitti
 * @since 2018.01.24
 */

public class SampleDiffUtilSmartRecyclerAdapter
    extends DiffUtilSmartRecyclerAdapter
{

  private static class TextDiffUtilCallback
      extends SmartDiffUtilCallback
  {

    public TextDiffUtilCallback(
        List<? extends SmartRecyclerViewWrapper<?>> oldWrappers,
        List<? extends SmartRecyclerViewWrapper<?>> newWrappers)
    {
      super(oldWrappers, newWrappers);
    }

    @Override
    protected Object getChangePayloadCustom(int oldItemPosition, int newItemPosition, Object oldBusinessObject,
        Object newBusinessObject)
    {
      return SmartDiffUtilCallback.ITEM_CHANGED_PAYLOAD;
    }
  }

  private SmartDiffUtilCallback smartDiffUtilCallback;

  public SampleDiffUtilSmartRecyclerAdapter(Context context)
  {
    super(context);
  }

  @Override
  public boolean onBindViewHolderCustom(SmartRecyclerAttributes holder, int position, Object payload)
  {
    return false;
  }

  @Override
  public <T extends SmartDiffUtilCallback> T getDiffUtilCallback(
      List<? extends SmartRecyclerViewWrapper<?>> oldWrappers,
      List<? extends SmartRecyclerViewWrapper<?>> newWrappers)
  {
    return (T) new TextDiffUtilCallback(oldWrappers, newWrappers);
  }
}

File: sample/src/main/java/com/smartnsoft/smartrecyclerviewsample/adapter/SampleDiffUtilSmartRecyclerAdapter.kt
package com.smartnsoft.smartrecyclerviewsample.adapter

import android.content.Context

import com.smartnsoft.recyclerview.adapter.DiffUtilSmartRecyclerAdapter
import com.smartnsoft.recyclerview.attributes.SmartRecyclerAttributes
import com.smartnsoft.recyclerview.wrapper.SmartRecyclerViewWrapper

/**
 * A simple implementation of the DiffUtilSmartRecyclerAdapter
 *
 * @author Adrien Vitti
 * @since 2018.01.24
 */

class SampleDiffUtilSmartRecyclerAdapter(context: Context)
  : DiffUtilSmartRecyclerAdapter(context)
{

  class ListDiffUtilCallback(oldWrappers: List<SmartRecyclerViewWrapper<*>>, newWrappers: List<SmartRecyclerViewWrapper<*>>)
    : DiffUtilSmartRecyclerAdapter.SmartDiffUtilCallback(oldWrappers, newWrappers)
  {

    override fun getChangePayloadCustom(oldItemPosition: Int, newItemPosition: Int, oldBusinessObject: Any?, newBusinessObject: Any?): Any? =
        SmartDiffUtilCallback.ITEM_CHANGED_PAYLOAD

  }

  override fun <T : SmartDiffUtilCallback?> getDiffUtilCallback(oldWrappers: List<SmartRecyclerViewWrapper<*>>, newWrappers: List<SmartRecyclerViewWrapper<*>>): T =
      ListDiffUtilCallback(oldWrappers, newWrappers) as T

  override fun onBindViewHolderCustom(holder: SmartRecyclerAttributes<*>?, position: Int, payload: Any?): Boolean =
      false

}
